File: src/main/com/stockcomp/configuration/SecurityConfiguration.java
package com.stockcomp.configuration;

import com.stockcomp.user.service.DefaultUserService;
import com.stockcomp.user.service.UserService;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.oauth2.server.resource.OAuth2ResourceServerConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.oauth2.jwt.JwtDecoder;
import org.springframework.security.oauth2.jwt.NimbusJwtDecoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.web.cors.CorsConfiguration;

import java.util.List;

@Configuration
@EnableWebSecurity
public class SecurityConfiguration {

    private final UserService userService;

    @Value("${admin.password}")
    private String adminPassword;

    @Value("${admin.email}")
    private String adminEmail;

    @Value("${spring.security.oauth2.resourceserver.jwt.jwk-set-uri}")
    String jwkSetUri;

    public SecurityConfiguration(DefaultUserService userService) {
        this.userService = userService;
    }

    @Bean
    public SecurityFilterChain configure(HttpSecurity httpSecurity) throws Exception {
        httpSecurity
                .cors().configurationSource(request -> createCorsConfiguration())
                .and()
                .csrf().disable()
                .authorizeHttpRequests(authorize -> authorize
                        .requestMatchers("/task/*", "/actuator/*","/user/*",
                                "/contest-operations/*", "/swagger-ui/*", "/swagger-resources/**", "/v2/api-docs")
                        .permitAll()
                        .anyRequest().authenticated()
                ).oauth2ResourceServer(OAuth2ResourceServerConfigurer::jwt)
                .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS);

        return httpSecurity.build();
    }

    @Bean
    JwtDecoder jwtDecoder() {
        return NimbusJwtDecoder.withJwkSetUri(this.jwkSetUri).build();
    }

    private CorsConfiguration createCorsConfiguration() {
        var cors = new CorsConfiguration();
        cors.addAllowedOrigin("http://localhost:8000");
        cors.addAllowedOrigin("http://localhost:3000");
        cors.addAllowedOrigin("http://stockclient-service:80");
        cors.addAllowedOrigin("http://localhost:80");
        cors.addAllowedOrigin("http://stockcompclient.io");
        cors.setAllowCredentials(true);
        cors.setAllowedMethods(List.of("GET", "POST", "PUT", "DELETE", "OPTIONS"));
        cors.setAllowedHeaders(List.of("Origin", "Access-Control-Allow-Origin", "Access-Control-Expose-Headers",
                "Content-Type", "Accept", "Authorization", "Origin,Accept", "X-Requested-With",
                "Access-Control-Request-Method", "Access-Control-Request-Headers", "Range"));
        cors.setExposedHeaders(List.of("Content-Range"));

        return cors;
    }

    @PostConstruct
    private void createAdminUser() {
        var adminUsername = "admin";
        var existingAdmin = userService.findUserByUsername(adminUsername);
        if (existingAdmin == null) {
            /*
            userService.signUpUser(
                    new SignUpRequest(adminUsername, adminPassword, adminEmail, Role.ADMIN));

             */
        }
    }
}

File: src/main/kotlin/com/stockcomp/configuration/SecurityConfiguration.kt
package com.stockcomp.configuration

import com.stockcomp.user.UserServiceExternal
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.JwtDecoder
import org.springframework.security.oauth2.jwt.NimbusJwtDecoder
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter
import org.springframework.security.web.SecurityFilterChain
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
@EnableMethodSecurity
class SecurityConfiguration(
    private val userService: UserServiceExternal,
) {
    @Value("\${spring.security.oauth2.resourceserver.jwt.jwk-set-uri}")
    var jwkSetUri: String? = null

    @Bean
    fun filterChain(httpSecurity: HttpSecurity): SecurityFilterChain =
        httpSecurity
            .cors { it.configurationSource(createCorsConfiguration()) }
            .csrf { it.disable() }
            .authorizeHttpRequests {
                it
                    .requestMatchers(
                        "/task/*",
                        "/actuator/*",
                        "/user/*",
                        "/contest-operations/*",
                        "/swagger-ui/*",
                        "/swagger-resources/**",
                        "/v2/api-docs",
                    ).permitAll()
                    .anyRequest()
                    .authenticated()
            }.oauth2ResourceServer { oauth2 ->
                oauth2.jwt { jwt -> jwt.jwtAuthenticationConverter(jwtAuthenticationConverter()) }
            }.sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
            .build()

    @Bean
    fun jwtDecoder(): JwtDecoder = NimbusJwtDecoder.withJwkSetUri(jwkSetUri).build()

    private fun createCorsConfiguration(): CorsConfigurationSource {
        val courseConfiguration =
            CorsConfiguration().apply {
                addAllowedOrigin("http://localhost:8000")
                addAllowedOrigin("http://localhost:3000")
                addAllowedOrigin("http://stockclient-service:80")
                addAllowedOrigin("http://localhost:80")
                allowCredentials = true
                allowedMethods = listOf("GET", "POST", "PUT", "DELETE", "OPTIONS")
                allowedHeaders =
                    listOf(
                        "Origin",
                        "Access-Control-Allow-Origin",
                        "Access-Control-Expose-Headers",
                        "Content-Type",
                        "Accept",
                        "Authorization",
                        "Origin,Accept",
                        "X-Requested-With",
                        "Access-Control-Request-Method",
                        "Access-Control-Request-Headers",
                        "Range",
                    )
                exposedHeaders = listOf("Content-Range")
            }
        return UrlBasedCorsConfigurationSource().also {
            it.registerCorsConfiguration("/**", courseConfiguration)
        }
    }

    private fun jwtAuthenticationConverter(): JwtAuthenticationConverter {
        val converter = JwtAuthenticationConverter()
        converter.setJwtGrantedAuthoritiesConverter { jwt ->
            val email = jwt.claims["email"] as String
            val userRole = userService.getUserRole(email)
            listOf(SimpleGrantedAuthority("ROLE_$userRole"))
        }
        return converter
    }
}
